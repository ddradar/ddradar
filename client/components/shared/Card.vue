<template>
  <section>
    <b-collapse class="card" animation="slide" :aria-id="title">
      <template v-slot:trigger="props">
        <header
          class="card-header"
          :class="type"
          role="button"
          :aria-controls="title"
        >
          <h1 class="card-header-title">{{ title }}</h1>
          <a class="card-header-icon">
            <b-icon :icon="props.open ? 'menu-down' : 'menu-up'" />
          </a>
        </header>
      </template>
      <slot />
    </b-collapse>
  </section>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'nuxt-property-decorator'

@Component
export default class CardComponent extends Vue {
  @Prop({ required: true, type: String })
  readonly title!: string

  @Prop({ required: true, type: String })
  readonly type!: string
}
</script>

<style lang="scss" scoped>
@import '~/assets/css/styles.scss';

.card {
  .card-header {
    @each $type, $color in $colors {
      &.is-#{$type} {
        background-color: nth($color, 1);
        .card-header-title,
        .card-header-icon {
          color: nth($color, 2);
        }
      }
    }
  }
}
</style>
