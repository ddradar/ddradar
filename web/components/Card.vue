<template>
  <o-collapse
    v-if="collapsible"
    class="card"
    animation="slide"
    :aria-id="title"
    :open="open"
  >
    <template #trigger>
      <header
        class="card-header"
        :class="color"
        role="button"
        :aria-controls="title"
      >
        <div class="card-header-title">{{ title }}</div>
        <a class="card-header-icon"><o-icon :icon="icon" /></a>
      </header>
    </template>
    <slot />
  </o-collapse>
  <div v-else class="card" animation="slide" :aria-id="title">
    <header class="card-header" :class="color" :aria-controls="title">
      <div class="card-header-title">{{ title }}</div>
    </header>
    <slot />
  </div>
</template>

<script lang="ts" setup>
import { computed } from 'vue'

export interface CardProps {
  /** Card title */
  title: string
  /** Card color */
  variant: string
  /** Collapsible card or not */
  collapsible?: boolean
  /** Card is opened or not (only affects collapsible card) */
  open?: boolean
}

const props = withDefaults(defineProps<CardProps>(), {
  collapsible: false,
  open: false,
})
const icon = computed(() => (props.open ? 'menu-down' : 'menu-up'))
const color = computed(() => `is-${props.variant}`)
</script>

<style lang="scss" scoped>
@import '~/assets/css/_colors';
.card {
  .card-header {
    @each $type, $color in $colors {
      &.is-#{$type} {
        background-color: nth($color, 1);
        .card-header-title,
        .card-header-icon {
          color: nth($color, 2);
        }
      }
    }
  }
}
</style>
