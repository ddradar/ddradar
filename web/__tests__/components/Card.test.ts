import Oruga from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'
import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'

import Card from '~/components/Card.vue'

describe('components/Card.vue', () => {
  test.each([
    { title: 'title', variant: 'primary' },
    { title: 'title', variant: 'danger', collapsible: true, open: false },
    { title: 'title', variant: 'info', collapsible: true, open: true },
  ])('(%o) snapshot test', props => {
    const wrapper = mount(Card, {
      props,
      global: { plugins: [[Oruga, bulmaConfig]] },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
