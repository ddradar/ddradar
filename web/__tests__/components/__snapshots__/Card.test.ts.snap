// Vitest Snapshot v1

exports[`components/Card.vue > ({ title: 'title', variant: 'danger', collapsible: true, open: false }) snapshot test 1`] = `
<div
  class="o-clps card"
  data-v-f883d62a=""
>
  <div
    class="o-clps__trigger"
  >
    <header
      aria-controls="title"
      class="card-header is-danger"
      data-v-f883d62a=""
      role="button"
    >
      <div
        class="card-header-title"
        data-v-f883d62a=""
      >
        title
      </div>
      <a
        class="card-header-icon"
        data-v-f883d62a=""
      >
        <span
          class="icon"
          data-v-f883d62a=""
        >
          <i
            class="mdi mdi-menu-down mdi-24px"
          />
        </span>
      </a>
    </header>
  </div>
  <transition-stub>
    <div
      class="o-clps__content"
      id="title"
      style="display: none;"
    >
      
      
    </div>
  </transition-stub>
</div>
`;

exports[`components/Card.vue > ({ title: 'title', variant: 'info', collapsible: true, open: true }) snapshot test 1`] = `
<div
  class="o-clps card"
  data-v-f883d62a=""
>
  <div
    class="o-clps__trigger"
  >
    <header
      aria-controls="title"
      class="card-header is-info"
      data-v-f883d62a=""
      role="button"
    >
      <div
        class="card-header-title"
        data-v-f883d62a=""
      >
        title
      </div>
      <a
        class="card-header-icon"
        data-v-f883d62a=""
      >
        <span
          class="icon"
          data-v-f883d62a=""
        >
          <i
            class="mdi mdi-menu-up mdi-24px"
          />
        </span>
      </a>
    </header>
  </div>
  <transition-stub>
    <div
      class="o-clps__content"
      id="title"
    >
      
      
    </div>
  </transition-stub>
</div>
`;

exports[`components/Card.vue > ({ title: 'title', variant: 'primary' }) snapshot test 1`] = `
<div
  animation="slide"
  aria-id="title"
  class="card"
  data-v-f883d62a=""
>
  <header
    aria-controls="title"
    class="card-header is-primary"
    data-v-f883d62a=""
  >
    <div
      class="card-header-title"
      data-v-f883d62a=""
    >
      title
    </div>
  </header>
  
  
</div>
`;
